signals.c

```diff
@@ -0,0 +1,95 @@
+/*
+  Not part of Grbl. KeyMe specific.
+  
+  Signal is one layer of abstraction above adc.h and adc.c.
+
+  ADC values are read in specific time intervals, filtered and
+  stored in the appropriate arrays.
+*/
+
+#include "signals.h"
+#include "adc.h"
+#include "systick.h"
+
+#define N_FILTER 3
+#define SIGNALS_DEFAULT_INTERVAL 1000 
+#define x(i) adc_samples_x[FORCE_VALUE_INDEX][i]
+
+// Unfiltered ADC readings
+int16_t adc_samples_x[VOLTAGE_SENSOR_COUNT][N_FILTER + 1]; 
```

Comment:
To make it easier to compare the code of the filters to the math of the filters.

Our filter is: y[n] = 0.25 * (x[n] + 2 * x[n - 1] + x[n - 2]). We use three samples, so N_FILTER = 3. N refers to n in the equation.

So the code can be written, for example, as:

`adc_sample[N_FILTER] = 0.25 * (X_BUF[N_FILTER] + 2 * X_BUF[N_FILTER - 1] + X_BUF[N_FILTER - 2])`

If we don't keep the extra non-used value in the array, we need to offset everything in our code with 1 from the math of the filter:

`adc_sample[N_FILTER - 1] = 0.25 * (X_BUF[N_FILTER - 1] + 2 * X_BUF[N_FILTER - 2] + X_BUF[N_FILTER - 3])`

 which adds mental overhead, especially with more advanced filtering and if we implement different filters for different ADCs.